reactive/containerd.py

```diff
@@ -257,21 +300,75 @@ def publish_config():
     )
 
 
-@when_not('containerd.br_netfilter.enabled')
-def enable_br_netfilter_module():
+@when('endpoint.docker-registry.ready')
+@when_not('containerd.registry.configured')
+def configure_registry():
     """
-    Enable br_netfilter to work
-    around https://github.com/kubernetes/kubernetes/issues/21613
+    Add docker registry config when present.
 
-    :returns: None
+    :return: None
     """
-    try:
-        modprobe('br_netfilter', persist=True)
-    except Exception:
-        log(traceback.format_exc())
-        if host.is_container():
-            log('LXD detected, ignoring failure to load br_netfilter')
-        else:
-            log('LXD not detected, will retry loading br_netfilter')
-            return
-    set_state('containerd.br_netfilter.enabled')
+    registry = endpoint_from_flag('endpoint.docker-registry.ready')
+
+    # Handle TLS data
+    cert_subdir = registry.registry_netloc
+    if registry.has_tls():
+        # Ensure the CA that signed our registry cert is trusted.
+        install_ca_cert(registry.tls_ca, name='juju-docker-registry')
+
+    manage_registry_certs(cert_subdir, remove=False)
+    # todo: add certs to config https://github.com/containerd/cri/blob/master/docs/registry.md
+
+    # Handle auth data.
+    if registry.has_auth_basic():
+        #  Set docker registry with auth.
+        docker_registry = {
+            "url": registry.registry_netloc,
+            "username": registry.basic_user,
+            "password": registry.basic_password
+        }
+    else:
+        #  Set docker registry without auth.
+        docker_registry = {"url": registry.registry_netloc}
+
+    DB.set('registry', docker_registry)
+
+    set_state('config.changed')
+    set_state('containerd.registry.configured')
+
+
+@when('endpoint.docker-registry.changed',
+      'containerd.registry.configured')
+def reconfigure_registry():
+    """
+    Signal to update the registry config when something changes.
+
+    :return: None
+    """
+    remove_state('containerd.registry.configured')
+
+
+@when('containerd.registry.configured')
+@when_not('endpoint.docker-registry.joined')
+def remove_registry():
+    """
+    Remove registry config when the registry is no longer present.
+
+    :return: None
+    """
+    docker_registry = DB.get('registry', None)
+
+    if docker_registry:
+        # Remove from DB.
+        DB.unset('registry')
```

Comment:
You may want to flush after this: https://github.com/juju/charm-helpers/blob/master/charmhelpers/core/unitdata.py#L375

A flush/commit will normally happen at the end of a hook, but i *think* if you have multiple handlers queued up, your `registry` value will still be set even after this `unset` until the hook exits (@johnsca is that right?).

You're probably fine since this is gated on a `@not joined`, but if anything else is watching for `containerd.registry.configured` and doesn't realize the registry has departed, you might be in trouble.

Reply:
unitdata uses a singleton `Storage` object, so will have a single connection sharing the current transaction.  So any reads should see the updated (or in this case removed) value.  Including an explicit flush should be ok, though.